### PR TITLE
Docker-Haproxy TLS

### PR DESCRIPTION
**Pull request by VariableDeclared**
Added ssl verify none to server_ops, also changed
http mode to change to tcp when SSL is configured.

Updated README.

### CODE REVIEW

**Comment by VariableDeclared** [comment]
Don't Merge.

**Comment by VariableDeclared** [comment]
[#LP1815962](https://bugs.launchpad.net/layer-docker-registry/+bug/1815962)

File: README.md [hunk 1]
```diff
@@ -58,19 +58,17 @@ juju config docker-registry \
...
-Any registry SSL configuration must be removed before creating the proxy
-relation. If SSL is desired in a proxied environment, the administrator must
-ensure certificates used by the proxy are configured on the appropriate target
-units.
+#### TLS/SSL
+TLS is supported between `docker-registry` and `haproxy`.
+Although it is not yet supported with easyrsa, in order implement TLS you will need to remove any easyrsa relation on `docker-registry` then manually supply certs to [`haproxy`](https://bazaar.launchpad.net/~haproxy-team/charm-haproxy/trunk/view/head:/config.yaml) and [`docker-registry`](https://github.com/CanonicalLtd/docker-registry-charm/blob/master/config.yaml).
 
-A proxied registry environment can be deployed as follows:
+These certificates used should come from a CA trusted by default in Ubuntu.
```
Comment (kwmonroe) [inline README.md]: `s/These/The`

File: reactive/docker_registry.py [hunk 2]
```diff
@@ -246,13 +246,20 @@ def update_reverseproxy_config():
     #  https://bugs.launchpad.net/layer-docker-registry/+bug/1815459
     common_opts = "check inter 2000 rise 2 fall 5 maxconn 4096"
     is_primary = True
+    tls_ops = ""
```
Comment (kwmonroe) [inline reactive/docker_registry.py]: Super knit-picky, but can you call this `tls_opts` vs `_ops`?  That's consistent with `common_opts` and `server_opts` in this function.

File: reactive/docker_registry.py [hunk 3]
```diff
@@ -246,13 +246,20 @@ def update_reverseproxy_config():
...
+    if (
+        is_flag_set('config.set.tls-cert-path') and
+        is_flag_set('config.set.tls-key-path')
+    ):
+        tls_ops = "ssl verify none"
```
Comment (kwmonroe) [inline reactive/docker_registry.py]: I think we want `ssl verify required` here.  We've let users know valid certs are required in this setup, and so ssl health checks should be used/work.  We'd want `none` for "SSL passthru", but we're not supporting that here.

File: README.md [hunk 4]
```diff
@@ -58,19 +58,17 @@ juju config docker-registry \
 
 ### Proxied Registry
 
-This charm supports an `http` proxy relation that allows operators to
+This charm supports `http` and `tcp` proxy relation that allows operators to
```
Comment (kwmonroe) [inline README.md]: I don't think this is true.  This PR is adding support for registry<->proxy via TLS iff valid certs are configured on both ends.  IOW, this is "SSL termination" in haproxy speak.  We are not supporting "SSL passthru" here, which is what would be implied by a `tcp` relation.

Also I think the original messaging was more that the [http interface](https://github.com/CanonicalLtd/docker-registry-charm/blob/master/metadata.yaml#L17) is used by the charm, versus any implication of how haproxy would communicate with the registry.

File: README.md (hunk 1 above)
Comment (VariableDeclared) [reply README.md]: whoops

File: README.md [hunk 5]
```diff
@@ -58,19 +58,17 @@ juju config docker-registry \
...
-units.
+#### TLS/SSL
+TLS is supported between `docker-registry` and `haproxy`.
+Although it is not yet supported with easyrsa, in order implement TLS you will need to remove any easyrsa relation on `docker-registry` then manually supply certs to [`haproxy`](https://bazaar.launchpad.net/~haproxy-team/charm-haproxy/trunk/view/head:/config.yaml) and [`docker-registry`](https://github.com/CanonicalLtd/docker-registry-charm/blob/master/config.yaml).
 
-A proxied registry environment can be deployed as follows:
+These certificates used should come from a CA trusted by default in Ubuntu.
 
-```bash
+```bashs
```
[the remark this answers is not in the document]
Comment (VariableDeclared) [reply README.md]: double whoops

File: reactive/docker_registry.py (hunk 2 above)
Comment (VariableDeclared) [reply reactive/docker_registry.py]: whoops, shoulda spotted that one

**Comment by VariableDeclared** [comment]
Going to test locally.

**Comment by VariableDeclared** [comment]
@kwmonroe  Tested the `verify required` but haproxy gives me this:

```
verify is enabled by default but no CA file specified. If you're running on a LAN where you're certain to trust the server's certificate, please set an explicit 'verify none' statement on the 'server' line, or use 'ssl-server-verify none' in the global section to disable server-side verifications by default.
```.

We can discuss.

**Comment by VariableDeclared** [comment]
Tried out locally and was able to push using insecure registries and self-signed certs.

**Pull request merged**